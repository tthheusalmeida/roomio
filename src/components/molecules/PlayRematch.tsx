import { mergeClassNames } from "@/utils/classNames";
import Button from "../atoms/Button";
import CountdownWithProgress from "../organisms/CountdownWithProgress";
import { useEffect, useState } from "react";

interface TitleProps {
  onFinish: Function;
}

export default function PlayRematch({ onFinish }: TitleProps) {
  const [showPrimary, setShowPrimary] = useState(true);
  const [showSecondary, setShowSecondary] = useState(true);

  const handleCountdownTime = () => {
    onFinish?.(false);
  };

  const handleClick = (value: boolean, hidePrimary: boolean) => {
    onFinish?.(value);

    if (hidePrimary) {
      setShowPrimary(false);
    } else {
      setShowSecondary(false);
    }
  };

  return (
    <div className="w-full flex flex-col gap-4 bg-gray-800 rounded-md p-4">
      <CountdownWithProgress
        initialMinutes={1}
        mode="both"
        height="h-2"
        color="bg-violet-500"
        showLabel
        onFinish={handleCountdownTime}
      />

      <div className="w-full text-center text-2xl mb-2">Rematch?</div>

      <div className="w-full flex gap-4">
        <Button
          variant="primary"
          className={mergeClassNames(
            "transition-all duration-500",
            !showPrimary ? "hidden" : "visible"
          )}
          onClick={() => handleClick(false, false)}
        >
          No
        </Button>

        <Button
          variant="secondary"
          className={mergeClassNames(
            "transition-all duration-500",
            !showSecondary ? "hidden" : "visible"
          )}
          onClick={() => handleClick(true, true)}
        >
          Yes
        </Button>
      </div>
    </div>
  );
}
