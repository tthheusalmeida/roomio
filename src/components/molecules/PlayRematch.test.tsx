import { render, screen, fireEvent } from "@testing-library/react";
import PlayRematch from "./PlayRematch";

// Mock do CountdownWithProgress
jest.mock("../organisms/CountdownWithProgress", () => {
  return ({ onFinish }: { onFinish?: () => void }) => (
    <div data-testid="mock-countdown">
      <button onClick={onFinish}>Finish Countdown</button>
    </div>
  );
});

describe("PlayRematch component", () => {
  it("renders title and buttons", () => {
    render(<PlayRematch onFinish={() => {}} />);
    expect(screen.getByText("Rematch?")).toBeInTheDocument();
    expect(screen.getByText("Yes").closest("button")).toBeInTheDocument();
    expect(screen.getByText("No").closest("button")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(<PlayRematch onFinish={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  it("calls onFinish with correct values clicking on NO button should hides YES button", async () => {
    const handleFinish = jest.fn();
    render(<PlayRematch onFinish={handleFinish} />);

    const yesButton = screen.getByText("Yes").closest("button")!;
    const noButton = screen.getByText("No").closest("button")!;

    fireEvent.click(noButton);
    expect(handleFinish).toHaveBeenCalledWith(false);

    const yesBtnUpdated = screen.getByText("Yes").closest("button")!;
    const noBtnUpdated = screen.getByText("No").closest("button")!;

    expect(noBtnUpdated).not.toHaveClass("hidden");
    expect(yesBtnUpdated).toHaveClass("hidden");
  });

  it("calls onFinish with correct values clicking on YES button should hides NO button", async () => {
    const handleFinish = jest.fn();
    render(<PlayRematch onFinish={handleFinish} />);

    const yesButton = screen.getByText("Yes").closest("button")!;
    const noButton = screen.getByText("No").closest("button")!;

    fireEvent.click(yesButton);
    expect(handleFinish).toHaveBeenCalledWith(true);

    const yesBtnUpdated = screen.getByText("Yes").closest("button")!;
    const noBtnUpdated = screen.getByText("No").closest("button")!;

    expect(noBtnUpdated).toHaveClass("hidden");
    expect(yesBtnUpdated).not.toHaveClass("hidden");
  });

  it("calls onFinish when countdown finishes", () => {
    const handleFinish = jest.fn();
    render(<PlayRematch onFinish={handleFinish} />);

    const finishButton = screen.getByText("Finish Countdown");
    fireEvent.click(finishButton);

    expect(handleFinish).toHaveBeenCalledWith(false);
  });
});
