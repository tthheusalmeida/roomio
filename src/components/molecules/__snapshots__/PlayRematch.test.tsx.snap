// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PlayRematch component matches snapshot 1`] = `
<div>
  <div
    class="w-full flex flex-col gap-4 bg-gray-800 rounded-md p-4"
  >
    <div
      data-testid="mock-countdown"
    >
      <button>
        Finish Countdown
      </button>
    </div>
    <div
      class="w-full text-center text-2xl mb-2"
    >
      Rematch?
    </div>
    <div
      class="w-full flex gap-4"
    >
      <button
        class="relative inline-flex justify-center items-center px-4 py-2 gap-4 text-sm font-medium cursor-pointer disabled:cursor-not-allowed rounded-sm w-full bg-violet-900 text-[var(--color-foreground)] hover:enabled:bg-violet-800 disabled:bg-violet-400 disabled:text-violet-50 transition-all duration-500 visible"
      >
        <span
          class="flex items-center gap-4 visible"
        >
          No
        </span>
      </button>
      <button
        class="relative inline-flex justify-center items-center px-4 py-2 gap-4 text-sm font-medium cursor-pointer disabled:cursor-not-allowed rounded-sm w-full bg-violet-100 text-violet-800 hover:enabled:bg-violet-200 disabled:bg-white disabled:text-violet-100 transition-all duration-500 visible"
      >
        <span
          class="flex items-center gap-4 visible"
        >
          Yes
        </span>
      </button>
    </div>
  </div>
</div>
`;
